// @splunk/react-time-range/Dialog/Presets.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _lodash = require("lodash");

var _i18n = require("@splunk/ui-utils/i18n");

var _time = require("@splunk/time-range-utils/time");

var _PresetLink = _interopRequireDefault(require("./PresetLink"));

var _PresetsStyles = require("./PresetsStyles");

var _Panel = _interopRequireDefault(require("./Panel"));

var _jsxFileName = "/builds/swp/ui-platform/projects/lib/search-components/react-time-range/src/Dialog/Presets.tsx";

const Presets = props => {
  const handleClick = (e, {
    earliest,
    latest
  }) => {
    props.onRequestParseEarliest(earliest, props.panelId);
    props.onRequestParseLatest(latest, props.panelId);
    props.onChange(e, {
      earliest,
      latest
    });
  };

  function renderPresets() {
    const realtime = [];
    const past = [];
    const latest = [];
    const other = [];
    props.presets.forEach(ps => {
      const parseEarliest = (0, _time.parseTimeString)(ps.earliest);
      const parseEarliestM = parseEarliest.modifiers[0];
      const parseLatest = (0, _time.parseTimeString)(ps.latest);

      const link = /*#__PURE__*/_react.default.createElement(_PresetLink.default, {
        key: `${ps.earliest} ${ps.latest}`,
        earliest: ps.earliest,
        latest: ps.latest,
        onClick: handleClick,
        label: ps.label,
        "data-test": ps.label,
        __self: this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 46,
          columnNumber: 17
        }
      }, (0, _i18n._)(ps.label));

      if (parseEarliest === false) {
        other.push(link);
      } else if (parseEarliest.type === 'realTime' || parseLatest.type === 'realTime') {
        realtime.push(link);
      } else if ((0, _time.isLatestNow)(ps.latest) && parseEarliestM && parseEarliestM.unit) {
        latest.push(link);
      } else if (parseEarliestM && (parseEarliestM.amount || parseEarliestM.snap)) {
        past.push(link);
      } else {
        other.push(link);
      }
    });
    return /*#__PURE__*/_react.default.createElement(_PresetsStyles.Styled, {
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 72,
        columnNumber: 13
      }
    }, !!realtime.length && /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledColumn, {
      "data-test": "real-time-column",
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 74,
        columnNumber: 21
      }
    }, /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledHeading, {
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 75,
        columnNumber: 25
      }
    }, (0, _i18n._)('Real-time')), realtime), !!past.length && /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledColumn, {
      "data-test": "relative-column",
      "data-test-column-id": "past",
      "data-divider": !!realtime.length,
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 80,
        columnNumber: 21
      }
    }, /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledHeading, {
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 85,
        columnNumber: 25
      }
    }, (0, _i18n._)('Relative')), past), !!latest.length && /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledColumn, {
      "data-test": "relative-column",
      "data-test-column-id": "latest",
      "data-divider": !!realtime.length && !past.length,
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 90,
        columnNumber: 21
      }
    }, /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledHeading, {
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 95,
        columnNumber: 25
      }
    }, !past.length && (0, _i18n._)('Relative')), latest), !!other.length && /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledColumn, {
      "data-test": "other-column",
      "data-divider": !!realtime.length || !!past.length || !!latest.length,
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 100,
        columnNumber: 21
      }
    }, /*#__PURE__*/_react.default.createElement(_PresetsStyles.StyledHeading, {
      __self: this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 104,
        columnNumber: 25
      }
    }, (0, _i18n._)('Other')), other));
  }

  return /*#__PURE__*/_react.default.createElement(_Panel.default, (0, _extends2.default)({
    title: (0, _i18n._)('Presets')
  }, props, {
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 113,
      columnNumber: 9
    }
  }), renderPresets());
};

Presets.defaultProps = {
  placeholder: 'Select...',
  onChange: _lodash.noop,
  onRequestParseEarliest: _lodash.noop,
  onRequestParseLatest: _lodash.noop
};
var _default = Presets;
exports.default = _default;
