// @splunk/react-time-range/Dialog/PresetLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Link = _interopRequireDefault(require("@splunk/react-ui/Link"));

var _stylesAndData = _interopRequireDefault(require("./stylesAndData"));

const _excluded = ["children", "earliest", "latest", "label", "onClick"];
var _jsxFileName = "/builds/swp/ui-platform/projects/lib/search-components/react-time-range/src/Dialog/PresetLink.tsx";

function PresetLink(_ref) {
  let {
    children,
    earliest,
    latest,
    label,
    onClick
  } = _ref,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  const handleClick = e => {
    onClick(e, {
      earliest,
      latest
    });
  };

  return /*#__PURE__*/_react.default.createElement(_Link.default, (0, _extends2.default)({
    style: {
      display: 'block'
    },
    label: label,
    onClick: handleClick
  }, (0, _stylesAndData.default)(rest), {
    __self: this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 23,
      columnNumber: 9
    }
  }), children);
}

var _default = PresetLink;
exports.default = _default;
