// @splunk/react-time-range/Dialog/RealTime.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _time = require("@splunk/time-range-utils/time");

var _lodash = require("lodash");

var _i18n = require("@splunk/ui-utils/i18n");

var _keyboard = require("@splunk/ui-utils/keyboard");

var _Button = _interopRequireDefault(require("@splunk/react-ui/Button"));

var _Message = _interopRequireDefault(require("@splunk/react-ui/Message"));

var _Number = _interopRequireDefault(require("@splunk/react-ui/Number"));

var _Select = _interopRequireDefault(require("@splunk/react-ui/Select"));

var _Text = _interopRequireDefault(require("@splunk/react-ui/Text"));

var _AbsoluteValue = _interopRequireDefault(require("./AbsoluteValue"));

var _Styles = require("./Styles");

var _Panel = _interopRequireDefault(require("./Panel"));

const _excluded = ["earliest", "open", "onChange", "onRequestParseEarliest", "parseEarliest", "panelId"];
var _jsxFileName = "/builds/swp/ui-platform/projects/lib/search-components/react-time-range/src/Dialog/RealTime.tsx";

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const RealTime = _ref => {
  let {
    earliest: earliestProp,
    open,
    onChange,
    onRequestParseEarliest,
    parseEarliest,
    panelId
  } = _ref,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  const fromNumberFromProps = (0, _time.getFromNumber)(earliestProp);
  const fromUnitFromProps = (0, _time.getFromUnit)(earliestProp);
  const [fromNumber, setFromNumber] = (0, _react.useState)(fromNumberFromProps);
  const [fromUnit, setFromUnit] = (0, _react.useState)(fromUnitFromProps);
  const [hasError, setHasError] = (0, _react.useState)();
  const prevOpenRef = (0, _react.useRef)();
  const prevEarliestRef = (0, _react.useRef)();
  (0, _react.useEffect)(() => {
    prevEarliestRef.current = earliestProp;
  }, [earliestProp]);
  (0, _react.useEffect)(() => {
    prevOpenRef.current = open;
  }, [open]);

  if (open && prevEarliestRef.current && (earliestProp !== prevEarliestRef.current || !prevOpenRef.current)) {
    if (hasError) {
      setHasError(false);
    }

    const needsEarliestUpdate = !prevOpenRef.current || fromNumberFromProps !== fromNumber || fromUnitFromProps !== fromUnit;

    if (needsEarliestUpdate) {
      onRequestParseEarliest((0, _time.getRealTimeEarliest)(fromNumber, fromUnit), panelId);
    }
  }

  const handleFromNumberChange = (e, {
    value
  }) => {
    setFromNumber(value);
    onRequestParseEarliest((0, _time.getRealTimeEarliest)(value, fromUnit), panelId);
  };

  const handleFromUnitChange = (e, {
    value
  }) => {
    setFromUnit(value);
    onRequestParseEarliest((0, _time.getRealTimeEarliest)(fromNumber, value), panelId);
  };

  const handleApply = e => {
    // eslint-disable-next-line no-restricted-globals
    if (!isFinite(fromNumber)) {
      setHasError(true);
      return;
    }

    const earliest = (0, _time.getRealTimeEarliest)(fromNumber, fromUnit);
    const latest = 'rt';
    onChange(e, {
      earliest,
      latest
    });
  };

  const handleKeyDown = e => {
    if ((0, _keyboard.keycode)(e) === 'enter') {
      handleApply(e);
    }
  };

  const earliest = (0, _time.removeRealTime)((0, _time.getRealTimeEarliest)(fromNumber, fromUnit));
  const earliestParseValid = parseEarliest && earliest === parseEarliest.time;
  return /*#__PURE__*/_react.default.createElement(_Panel.default, (0, _extends2.default)({
    title: (0, _i18n._)('Real-time'),
    panelId: panelId,
    open: open
  }, rest, {
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 114,
      columnNumber: 9
    }
  }), hasError && /*#__PURE__*/_react.default.createElement(_Styles.StyledError, {
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 116,
      columnNumber: 17
    }
  }, /*#__PURE__*/_react.default.createElement(_Message.default, {
    type: "error",
    "data-test": "error-message",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 117,
      columnNumber: 21
    }
  }, (0, _i18n._)('A number is required for earliest.'))), /*#__PURE__*/_react.default.createElement(_Styles.StyledCGWrapper, {
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 122,
      columnNumber: 13
    }
  }, /*#__PURE__*/_react.default.createElement(_Styles.FlexControlGroup, {
    controlsLayout: "none",
    label: (0, _i18n._)('Earliest:'),
    labelPosition: "top",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 123,
      columnNumber: 17
    }
  }, /*#__PURE__*/_react.default.createElement(_Styles.StyledFromControl, {
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 124,
      columnNumber: 21
    }
  }, /*#__PURE__*/_react.default.createElement(_Number.default, {
    onChange: handleFromNumberChange,
    onKeyDown: handleKeyDown,
    min: 1,
    roundTo: 0,
    value: fromNumber,
    style: {
      flex: 1
    },
    append: true,
    "data-test": "earliest-number",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 125,
      columnNumber: 25
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default, {
    value: fromUnit,
    onChange: handleFromUnitChange,
    style: {
      flex: '0 1 auto'
    },
    prepend: true,
    "data-test": "earliest-unit",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 135,
      columnNumber: 25
    }
  }, /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Seconds Ago'),
    value: "s",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 142,
      columnNumber: 29
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Minutes Ago'),
    value: "m",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 143,
      columnNumber: 29
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Hours Ago'),
    value: "h",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 144,
      columnNumber: 29
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Days Ago'),
    value: "d",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 145,
      columnNumber: 29
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Weeks Ago'),
    value: "w",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 146,
      columnNumber: 29
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Months Ago'),
    value: "mon",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 147,
      columnNumber: 29
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Quarters Ago'),
    value: "q",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 148,
      columnNumber: 29
    }
  }), /*#__PURE__*/_react.default.createElement(_Select.default.Option, {
    label: (0, _i18n._)('Years Ago'),
    value: "y",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 149,
      columnNumber: 29
    }
  }))), /*#__PURE__*/_react.default.createElement(_AbsoluteValue.default, {
    error: earliestParseValid ? parseEarliest.error : undefined,
    displayValue: earliestParseValid ? parseEarliest.displayValue : undefined,
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 152,
      columnNumber: 21
    }
  })), /*#__PURE__*/_react.default.createElement(_Styles.FlexControlGroup, {
    controlsLayout: "none",
    label: (0, _i18n._)('Latest:'),
    labelPosition: "top",
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 157,
      columnNumber: 17
    }
  }, /*#__PURE__*/_react.default.createElement(_Text.default, {
    value: "now",
    "data-test": "now",
    disabled: true,
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 158,
      columnNumber: 21
    }
  }))), /*#__PURE__*/_react.default.createElement(_Styles.StyledApplyContainer, {
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 161,
      columnNumber: 13
    }
  }, /*#__PURE__*/_react.default.createElement(_Button.default, {
    label: (0, _i18n._)('Apply'),
    "data-test": "apply",
    disabled: !earliestParseValid || !!(parseEarliest && parseEarliest.error),
    onClick: handleApply,
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 162,
      columnNumber: 17
    }
  })));
};

RealTime.defaultProps = {
  open: false,
  onChange: _lodash.noop,
  onRequestParseEarliest: _lodash.noop,
  onRequestParseLatest: _lodash.noop
};
var _default = RealTime;
exports.default = _default;
