// @splunk/react-time-range/Dialog/Styles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledFromControl = exports.StyledApplyContainer = exports.FlexControlGroup = exports.StyledCGWrapper = exports.StyledError = void 0;

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ControlGroup = _interopRequireDefault(require("@splunk/react-ui/ControlGroup"));

var _themes = require("@splunk/themes");

const StyledError = _styledComponents.default.div`
    padding: 0 ${_themes.variables.spacingLarge};
`;
exports.StyledError = StyledError;
const StyledCGWrapper = _styledComponents.default.div`
    display: flex;
    padding: ${_themes.variables.spacingSmall} ${_themes.variables.spacingLarge} 0;

    ${
/* sc-sel */
StyledError} + & {
        padding-top: 0;
    }
`;
exports.StyledCGWrapper = StyledCGWrapper;
const FlexControlGroup = (0, _styledComponents.default)(_ControlGroup.default)`
    & {
        flex: 1;
    }

    &:first-child {
        margin-top: 0;
        padding-right: ${_themes.variables.spacingLarge};
    }
`;
exports.FlexControlGroup = FlexControlGroup;
const StyledApplyContainer = _styledComponents.default.div`
    text-align: right;
    padding: 0 ${_themes.variables.spacingLarge} ${_themes.variables.spacingLarge};

    & > button {
        min-width: 100px;
    }
`;
exports.StyledApplyContainer = StyledApplyContainer;
const StyledFromControl = _styledComponents.default.div`
    display: flex;
    margin-bottom: ${_themes.variables.spacingSmall};
`;
exports.StyledFromControl = StyledFromControl;
