// @splunkdev/cloud-sdk/services/catalog/v2alpha2/models/GeneratedImportDatasetKind.js
"use strict";
// tslint:disable
/**
 * Copyright 2020 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * Metadata Catalog
 * With the Metadata Catalog you can create and manage knowledge objects such as datasets, fields, rules, actions, dashboards, and workflows.
 *
 * OpenAPI spec version: v2alpha2.6
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
exports.__esModule = true;
/**
 * The dataset kind.
 * @export
 */
var ImportDatasetKind;
(function (ImportDatasetKind) {
    ImportDatasetKind["Import"] = "import";
})(ImportDatasetKind = exports.ImportDatasetKind || (exports.ImportDatasetKind = {}));
function ImportDatasetKindFromJSON(json) {
    return json;
}
exports.ImportDatasetKindFromJSON = ImportDatasetKindFromJSON;
function ImportDatasetKindToJSON(value) {
    return value;
}
exports.ImportDatasetKindToJSON = ImportDatasetKindToJSON;
