// @splunkdev/cloud-sdk/services/catalog/v2beta1/models/GeneratedSplv1sinkDatasetKind.js
"use strict";
// tslint:disable
/**
 * Copyright 2020 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * Metadata Catalog service
 * With the Metadata Catalog in Splunk Cloud Services you can create and manage knowledge objects such as datasets, fields, rules, actions, dashboards, and workflows.
 *
 * OpenAPI spec version: v2beta1.4 (recommended default)
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
exports.__esModule = true;
/**
 * The dataset kind.
 * @export
 */
var Splv1sinkDatasetKind;
(function (Splv1sinkDatasetKind) {
    Splv1sinkDatasetKind["Splv1sink"] = "splv1sink";
})(Splv1sinkDatasetKind = exports.Splv1sinkDatasetKind || (exports.Splv1sinkDatasetKind = {}));
function Splv1sinkDatasetKindFromJSON(json) {
    return json;
}
exports.Splv1sinkDatasetKindFromJSON = Splv1sinkDatasetKindFromJSON;
function Splv1sinkDatasetKindToJSON(value) {
    return value;
}
exports.Splv1sinkDatasetKindToJSON = Splv1sinkDatasetKindToJSON;
