// @splunk/react-time-range/TimeAcceleratorPopup/BaseDropdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Dropdown = _interopRequireDefault(require("@splunk/react-ui/Dropdown"));

var _Menu = _interopRequireDefault(require("@splunk/react-ui/Menu"));

var _Button = _interopRequireDefault(require("@splunk/react-ui/Button"));

var _jsxFileName = "/builds/swp/ui-platform/projects/lib/search-components/react-time-range/src/TimeAcceleratorPopup/BaseDropdown.tsx";

const BaseDropdown = ({
  id: menuId,
  selected,
  append,
  prepend,
  menuItemIds,
  menuItems,
  onMenuItemClick
}) => {
  const baseMenuItems = menuItemIds.map(id => {
    return /*#__PURE__*/_react.default.createElement(_Menu.default.Item, {
      onClick: e => {
        onMenuItemClick(e, id);
      },
      selectable: true,
      selected: id === selected,
      key: id,
      value: id,
      __self: void 0,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 37,
        columnNumber: 13
      }
    }, menuItems[id]);
  });

  const toggle = /*#__PURE__*/_react.default.createElement(_Button.default, {
    isMenu: true,
    appearance: "toggle",
    append: append,
    prepend: prepend,
    label: menuItems[selected],
    "data-test": `time-accelerator-${menuId}`,
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 52,
      columnNumber: 9
    }
  });

  return /*#__PURE__*/_react.default.createElement(_Dropdown.default, {
    toggle: toggle,
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 63,
      columnNumber: 9
    }
  }, /*#__PURE__*/_react.default.createElement(_Menu.default, {
    __self: void 0,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 64,
      columnNumber: 13
    }
  }, baseMenuItems));
};

var _default = BaseDropdown;
exports.default = _default;
