// @splunkdev/cloud-sdk/services/streams/v2beta1/models/GeneratedSplCompileRequest.js
"use strict";
// tslint:disable
/**
 * Copyright 2020 Splunk, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"): you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * Data Stream Processing REST API
 * Use the Streams service to perform create, read, update, and delete (CRUD) operations on your data pipeline. The Streams service also has metrics and preview session endpoints and gives you full control over your data pipeline.
 *
 * OpenAPI spec version: v2beta1.4
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
exports.__esModule = true;
/**
 * @export
 */
var SplCompileRequestSyntaxEnum;
(function (SplCompileRequestSyntaxEnum) {
    SplCompileRequestSyntaxEnum["UPL"] = "UPL";
    SplCompileRequestSyntaxEnum["DSL"] = "DSL";
    SplCompileRequestSyntaxEnum["SPL"] = "SPL";
    SplCompileRequestSyntaxEnum["EVAL"] = "EVAL";
    SplCompileRequestSyntaxEnum["WHERE"] = "WHERE";
    SplCompileRequestSyntaxEnum["TIMECHART"] = "TIMECHART";
    SplCompileRequestSyntaxEnum["FIELDS"] = "FIELDS";
    SplCompileRequestSyntaxEnum["MVEXPAND"] = "MVEXPAND";
    SplCompileRequestSyntaxEnum["REX"] = "REX";
    SplCompileRequestSyntaxEnum["BIN"] = "BIN";
    SplCompileRequestSyntaxEnum["RENAME"] = "RENAME";
    SplCompileRequestSyntaxEnum["STATS"] = "STATS";
    SplCompileRequestSyntaxEnum["STATSBY"] = "STATS_BY";
    SplCompileRequestSyntaxEnum["SELECT"] = "SELECT";
    SplCompileRequestSyntaxEnum["EXPRESSION"] = "EXPRESSION";
    SplCompileRequestSyntaxEnum["FUNCTION"] = "FUNCTION";
    SplCompileRequestSyntaxEnum["LITERAL"] = "LITERAL";
    SplCompileRequestSyntaxEnum["UNKNOWN"] = "UNKNOWN";
})(SplCompileRequestSyntaxEnum = exports.SplCompileRequestSyntaxEnum || (exports.SplCompileRequestSyntaxEnum = {}));
